### ISSUE TITLE
what about logout?

### ISSUE THREAD

**Issue by gabrielmancini** [dwyl/elixir-auth-microsoft#29]
Hello... great lib btw.

what about logout? there is some gotcha to block an ad user and the user logout the phoenix app?

thanks

**Comment by LuchoTurtle** [comment]
Hey @gabrielmancini !
Thanks for opening the issue :)
I'm afraid I don't quite understand your phrasing. Do you want a function that logs out an AD user from your Phoenix App?

**Comment by gabrielmancini** [comment]
yes, when i block, or delete or revoke user session on active directory like this:

![ad](https://user-images.githubusercontent.com/144915/227509773-e983d31e-5e14-4189-8bdc-a3966a3249dc.png)

i want to the user logout from the app

**Comment by LuchoTurtle** [comment]
Although I think there's something we can do about logging the user out of the app conditionally by implementing a `logout()` function, I think the behaviour you want is currently unattainable through Azure.

The problem you've described is effectively an auth strategy downside to having **access tokens** and **refresh tokens**.
When an user is authorized, Azure AD issues these tokens for the resource. These last **1 hour** by default. You may silently reauthenticate the user by passing the refresh token whenever the access token expires after one hour.
However, if you want to *revoke* an access token during that time, the user will only effectively lose access to the resource after the expiration period is surpassed (because they can't refresh the token - it was revoked).

This issue is not related only to this package, it's generalized to packages that use this auth grant workflow. 
For example, [`msal-react`](https://www.npmjs.com/package/@azure/msal-react) (which is the official Microsoft package for React web applications to deal with authentication with Azure) also has this "problem".

Labelling this as a "problem" is a debate on itself. You can certainly reduce the expiration window to 30 minutes so the user has to silently re-authenticate and have a new access/refresh token pair but this will spark a debate on scalability/latency/etc...

Azure is developing a thing called [Continuous Access Evaluation](https://learn.microsoft.com/en-us/azure/active-directory/conditional-access/concept-continuous-access-evaluation) to tackle this subject, where they evaluate the token's validity near-realtime. But this is only available on Exchange, Teams, and SharePoint Online resources, not all Azure resources.

This library makes it easier for you to follow the [OAuth 2.0 authorization code flow](https://learn.microsoft.com/en-us/azure/active-directory/develop/v2-oauth2-auth-code-flow) (recommend reading this doc from Microsoft to understand how tokens work in this authorization flow).

If you want to learn more about Revoking user access on Azure AD, check https://learn.microsoft.com/en-us/azure/active-directory/enterprise-users/users-revoke-access#access-tokens-and-refresh-tokens.

# tl;dr

When you revoke user access in Azure AD, the user won't immediately lose access to your resource. You'll have to wait until the access token expiration time is reached and they no longer can refresh the token because their access was revoked. 

What can *possibly* be done is implementing a callback URL into this library so that whenever the user is not able to re-authenticate, logging the user out and redirecting them to the homepage of the Phoenix application.

**Comment by LuchoTurtle** [comment]
Hm, having a bit of trouble with this...

I'm getting an [`ErrorInsufficientPermissionsInAccessToken`](https://learn.microsoft.com/en-us/answers/questions/1084592/profileaccessdeniedexception-on-profile-picture-en) error when trying to fetch the profile with the token. 

Having read through https://learn.microsoft.com/en-us/answers/questions/880570/having-an-issue-with-oauth2, 
I've already checked...
- the delegated permissions of the app registration.
<img width="1090" alt="image" src="https://user-images.githubusercontent.com/17494745/230451490-24dc72e8-4e8e-46e6-9587-242a714d6f14.png">

- the contents of the `id_token` jwt .
<img width="1143" alt="image" src="https://user-images.githubusercontent.com/17494745/230451700-529e6f85-b622-4235-89d0-cc83ae80c629.png">

- and the supported account types.
<img width="737" alt="image" src="https://user-images.githubusercontent.com/17494745/230451768-388407ac-7062-49bb-a359-6001936fea1f.png">

But I keep stumbling upon this error trying to fetch the user profile...

I've inclusively refreshed the secret and have updated it accordingly but it doesn't seem to work :/

<img width="597" alt="image" src="https://user-images.githubusercontent.com/17494745/230451948-e3140a51-ad3e-4e98-bd74-0ce4802804a5.png">

**Comment by gabrielmancini** [comment]
i think this can be relevant to this question: https://learn.microsoft.com/en-us/azure/active-directory/develop/v2-protocols-oidc#send-a-sign-out-request

**Comment by LuchoTurtle** [comment]
#33 should address this.
For more information on how the logout process works, Microsoft has this document with the endpoint people need to be redirected to so they end their account session within Microsoft's identity server -> https://learn.microsoft.com/en-us/azure/active-directory/develop/v2-protocols-oidc#send-a-sign-out-request.
